qiskit/algorithms/state_fidelities/base_state_fidelity.py

```diff
@@ -0,0 +1,292 @@
+            if len(values) > 0 and not isinstance(values[0], list):
+                values = [values]
+            return values
+
+    def _check_qubits_match(self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit) -> None:
+        """
+        Checks that the number of qubits of 2 circuits matches.
+        Args:
+            circuit_1: (Parametrized) quantum circuit
+            circuit_2: (Parametrized) quantum circuit
+
+        Raises:
+            ValueError: when ``circuit_1`` and ``circuit_2`` don't have the
+            same number of qubits.
+        """
+
+        if circuit_1.num_qubits != circuit_2.num_qubits:
+            raise ValueError(
+                f"The number of qubits for the first circuit ({circuit_1.num_qubits}) "
+                f"and second circuit ({circuit_2.num_qubits}) do not coincide."
+            )
+
+    @abstractmethod
+    def create_fidelity_circuit(
+        self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit
+    ) -> QuantumCircuit:
+        """
+        Implementation-dependent method to create a fidelity circuit
+        from 2 circuit inputs.
+        Args:
+            circuit_1: (Parametrized) quantum circuit
+            circuit_2: (Parametrized) quantum circuit
+
+        Returns:
+            The fidelity quantum circuit corresponding to ``circuit_1`` and ``circuit_2``.
+        """
+        raise NotImplementedError
+
+    def _construct_circuits(
+        self,
+        circuits_1: Sequence[QuantumCircuit],
+        circuits_2: Sequence[QuantumCircuit],
+    ) -> Sequence[QuantumCircuit]:
+        """
+        Construct the list of fidelity circuits to be evaluated.
+        These circuits represent the state overlap between pairs of input circuits,
+        and their construction depends on the fidelity method implementations.
+
+        Args:
+            circuits_1: (Parametrized) quantum circuits.
+            circuits_2: (Parametrized) quantum circuits.
+
+        Returns:
+            List of constructed fidelity circuits.
+
+        Raises:
+            ValueError: if the length of the input circuit lists doesn't match.
+        """
+
+        if not len(circuits_1) == len(circuits_2):
+            raise ValueError(
+                f"The length of the first circuit list({len(circuits_1)}) "
+                f"and second circuit list ({len(circuits_2)}) does not coincide."
+            )
+
+        circuits = []
+        for (circuit_1, circuit_2) in zip(circuits_1, circuits_2):
+
+            # TODO: improve caching, what if the circuit is modified without changing the id?
+            circuit = self._circuit_cache.get((id(circuit_1), id(circuit_2)))
+
+            if circuit is not None:
+                circuits.append(circuit)
+            else:
+                self._check_qubits_match(circuit_1, circuit_2)
+
+                # re-parametrize input circuits
+                # TODO: make smarter checks to avoid unnecesary reparametrizations
```

Comment:
I am imagining that these 2 TODOs are for potential future work and are not being addressed at this time.